From a69c849b8fb0d450a44f26ec42adc127ea1fccda Mon Sep 17 00:00:00 2001
From: Hongxu Jia <hongxu.jia@windriver.com>
Date: Tue, 7 Apr 2015 16:08:37 +0800
Subject: [PATCH] build initrd for lvm

While using lvm fs, it must be booted from initrd to mount
the root fs, so we build the initramfs by invoking the script
from lvm2

Upstream-Status: inappropriate [wr specific]

Signed-off-by: Hongxu Jia <hongxu.jia@windriver.com>
---
 pyanaconda/dispatch.py         |  2 ++
 pyanaconda/installclass.py     |  1 +
 pyanaconda/storage/__init__.py | 14 ++++++++++++++
 3 files changed, 17 insertions(+)

diff --git a/pyanaconda/dispatch.py b/pyanaconda/dispatch.py
index 5d47828..466e262 100644
--- a/pyanaconda/dispatch.py
+++ b/pyanaconda/dispatch.py
@@ -36,6 +36,7 @@ from packages import setFileCons
 from storage import storageInitialize
 from storage import storageComplete
 from storage.partitioning import doAutoPartition
+from storage import buildLvmInitrd
 from bootloader import writeBootloader
 from flags import flags
 from upgrade import upgradeMountFilesystems
@@ -283,6 +284,7 @@ class Dispatcher(object):
         self.add_step("postinstallconfig", doPostInstall)
         self.add_step("writeconfig", writeConfiguration)
         self.add_step("firstboot", firstbootConfiguration)
+        self.add_step("buildlvminitrd", buildLvmInitrd)
         self.add_step("instbootloader", writeBootloader)
         self.add_step("reipl", doReIPL)
         self.add_step("writeksconfig", writeKSConfiguration)
diff --git a/pyanaconda/installclass.py b/pyanaconda/installclass.py
index 90a6710..3a577e1 100644
--- a/pyanaconda/installclass.py
+++ b/pyanaconda/installclass.py
@@ -112,6 +112,7 @@ class BaseInstallClass(object):
                  "postinstallconfig",
 		 "writeconfig",
                  "firstboot",
+                 "buildlvminitrd",
 		 "instbootloader",
                  "dopostaction",
                  "postscripts",
diff --git a/pyanaconda/storage/__init__.py b/pyanaconda/storage/__init__.py
index 3c936a2..a193643 100644
--- a/pyanaconda/storage/__init__.py
+++ b/pyanaconda/storage/__init__.py
@@ -64,6 +64,20 @@ _ = lambda x: gettext.ldgettext("anaconda", x)
 import logging
 log = logging.getLogger("storage")
 
+def buildLvmInitrd(anaconda):
+    from glob import glob
+    storage = anaconda.storage
+    log.info('/ is %s type' % anaconda.storage.mountpoints['/'].type)
+    if anaconda.storage.mountpoints['/'].type != 'lvmlv':
+        return
+
+    for f in glob("/boot/bzImage-*"):
+        initrd = "/boot/initrd-" + f.replace("/boot/bzImage-", "")
+        log.info("build_lvm_initramfs.sh %s" % initrd)
+        iutil.execWithRedirect("build_lvm_initramfs.sh", [initrd],
+                               stdout = "/dev/tty5", stderr="/dev/tty5",
+                               root = anaconda.rootPath)
+
 def storageInitialize(anaconda):
     storage = anaconda.storage
 
-- 
1.9.1

